### ISSUE TITLE
have `cabal init` create Main.hs

### ISSUE THREAD

**Issue by cdepillabout** [haskell/cabal#2304]
I often use `cabal init` when creating a new package, often before I've even started writing any code.  Using `cabal-install` version 1.20.0.4, it doesn't seem like `cabal init` will create `Main.hs` for you, even if it doesn't already exist.

It would be nice `cabal init` created a short Main.hs for you, if it doesn't already exist.

Ideally, I'd like to be able to have the following workflow:

```sh
# make a new directory for my project
$ mkdir my-project && cd my-project

# initialize my sandbox
$ cabal sandbox init
...

# create the .cabal file
$ cabal init
...

# cabal init would have created a Main.hs like below
$ cat Main.hs
module Main where

main = putStrLn "Hello World!"

# now, you can immediately run `cabal install`
$ cabal install
...

# this produces an executable that you can immediately run
$ ./.cabal-sandbox/bin/my-project
Hello World!
$
```

If I were to send a pull request for this, would there be any interest in it?

**Comment by fujimura** [comment]
In my understanding, `cabal init` is a command to cabalise the exisiting project. So, generating files are a bit different from original purpose.

Aside from original purpose, I think if `cabal init` generates sufficient files to start new project, it'd be helpful for many developers.

**Comment by byorgey** [comment]
Yes, I don't think this really fits with the purpose of `cabal init`.  @cdepillabout , perhaps you would be interested in a project like https://github.com/fujimura/hi which seems to be more what you are looking for.

**Comment by cdepillabout** [comment]
@byorgey, you're right.  `hi` is a really nice project that could definitely be used for what I want, but it's a little "heavy" for just generating a `Main.hs` file.

I realize that `cabal init` was originally used for cabalizing an existing Haskell project, but I almost always end up using it before I actually write any Haskell code.  For my specific use case, it would be nice if it created a `Main.hs` for me.

Maybe I should seriously look into using `hi` to generate projects instead of `cabal init`.

**Comment by byorgey** [comment]
@cdepillabout , if `hi` is too heavyweight, you could just write a little script that calls `cabal init` and then copies a fresh `Main.hs` into the directory.

**Comment by cdepillabout** [comment]
That's true, I'll look into doing that.

**Comment by cgag** [comment]
Is the purpose really to cabalize existing non-cabal projects?  I might be wrong but I feel starting new projects with "cabal init" has to be way more common.  Creating a working project from "cabal init" seems like a very nice thing to have, especially for beginners.

**Comment by 23Skidoo** [comment]
Perhaps `init` should have two modes of operation. For empty directories it'd default to "new project" mode and for non-empty to "cabalize" mode.

**Comment by erikd** [comment]
I think @23Skidoo's ideas of two modes is a good. Personally, "cabalize" an existing project is by far my most common use case.

**Comment by bitemyapp** [comment]
Two modes...could be okay. I'm a little iffy on it. The overloading of commands is already getting to be a bit much.

I'm the opposite of @erikd - cabal init is usually my first command. It's possible that an official templating function could help here.

**Comment by cdepillabout** [comment]
@bitemyapp, I'm the same as you.  

As for a templating function, have you tried `hi` mentioned above?  It is what you're looking for?

What I'd like more than a templating function is just something that creates `Main.hs` so I can run `cabal run` (or `cabal install`) right away.  For my personal use-case it doesn't have to be a templating function.

**Comment by cgag** [comment]
I'm not actually sure why it would need two different modes.  While I think cabal init is probably most common use-case, this seems like an improvement for both uses.  When "cabalizing" a project, I imagine you'd already have a main that already exists, so this wouldn't affect things.  If you didn't, I imagine you'd need to create one as well.

Re: hi templating, I'd like it if cabal just did templating.  I know lein (Clojure) does it, and I believe it's being considered by Cargo (rust) as well.

**Comment by byorgey** [comment]
Well, perhaps you are all right, having `cabal init` just create a simple `Main.hs` if one does not already exist would be useful.  I will reopen this issue; pull requests welcome.

**Comment by cdepillabout** [comment]
This has been merged in #2483.  Thanks!
